File: src/server/Services/PlayerService.lua
local Knit = _G.Knit

local Players = game:GetService("Players")
local RunService = game:GetService("RunService")

local isStudio = RunService:IsStudio()
local global = require(Knit.Shared.globalData)
local RemoteEvent = require(Knit.Util.Remote.RemoteEvent)

local PlayerObject = require(Knit.Modules.playerObject)
local ProfileService = require(Knit.Modules.ProfileService)
local datastore = ProfileService.GetProfileStore("Data"..global.DATA_VERSION, global.DEFAULT_DATA)

local profiles = {}
local readyPlayers = {}
local readyEvent = RemoteEvent.new()
local updateEvent = RemoteEvent.new()

local PlayerService = Knit.CreateService{
    Name = "PlayerService";
    profiles = profiles;

    Client = {
        Ready = readyEvent;
        Update = updateEvent;
    }
}


function loadData(player)
    local profile

    if isStudio then
        profile = datastore.Mock:LoadProfileAsync("Player_"..player.UserId, "ForceLoad")
    else
        profile = datastore:LoadProfileAsync("Player_"..player.UserId, "ForceLoad")
    end

    if not profile then
        return player:Kick("Data unable to load. Contact devs.")
    end

    profile:Reconcile()
    profile:ListenToRelease(function()
        if player:IsDescendantOf(Players) then
            player:Kick("Data loaded from somewhere else.")
        end
    end)

    if not player:IsDescendantOf(Players) then
        return profile:Release()
    end

    return profile
end

function PlayerService.KnitInit()
    readyEvent:Connect(function(player)
        if table.find(readyPlayers, player) then return end
        table.insert(readyPlayers, player)

        local profile = loadData(player)
        if not profile then return end

        local playerProfile = PlayerObject.new(player, profile)
        profiles[player] = playerProfile
        readyEvent:Fire(player, playerProfile.data)
    end)

    Players.PlayerRemoving:Connect(function(player)
        local playerProfile = profiles[player]
        if playerProfile then
            playerProfile.profile:Release()
            table.remove(readyPlayers, table.find(readyPlayers, player))
        end
    end)
end

return PlayerService

File: src/server/services/playerService.ts
import { Players, RunService } from "@rbxts/services";
import { OnInit, Service } from "@rbxts/flamework";
import { DATA_VERSION, DEFAULT_DATA } from "shared/globalData";
import { server } from "shared/globalEvents";
import ProfileService from "@rbxts/profileservice";
import playerObject from "../modules/playerObject";

const isStudio = RunService.IsStudio();
const datastore = ProfileService.GetProfileStore(`Data${DATA_VERSION}`, DEFAULT_DATA);

export const profiles = new Map<Player, playerObject>();

function loadData(player: Player) {
	let profile;

	if (isStudio) {
		profile = datastore.Mock.LoadProfileAsync(`Player_${player.UserId}`, "ForceLoad");
	} else {
		profile = datastore.LoadProfileAsync(`Player_${player.UserId}`, "ForceLoad");
	}

	if (!profile) {
		return player.Kick("Data unable to load.");
	}

	profile.Reconcile();
	profile.ListenToRelease(() => {
		if (player.IsDescendantOf(Players)) {
			player.Kick("Data loaded from somewhere else.");
		}
	});

	if (!player.IsDescendantOf(Players)) {
		return profile.Release();
	}

	return profile;
}

@Service({
	loadOrder: 0,
})
export class playerService implements OnInit {
	onInit() {
		server.connect("ready", (player) => {
			if (profiles.has(player)) return;

			const profile = loadData(player);
			if (!profile) return;

			profiles.set(player, new playerObject(player, profile));
			server.stateUpdate(player, "dataReady", profile.Data);
		});

		Players.PlayerRemoving.Connect((player) => {
			const playerProfile = profiles.get(player);
			if (!playerProfile) return;

			playerProfile.profile.Release();
			profiles.delete(player);
		});
	}
}
